### ISSUE TITLE
Multiple outputs

### ISSUE THREAD

**Issue by atremblay** [GRAAL-Research/poutyne#29]
Would it be wise to support multiple output? At least a pass through? I have a model that returns two outputs (prediction and attention weights). I had to do a small wrapper around the loss function as such

```python
    def loss_function(y_pred, y_true):
        output, word_attention = y_pred
        return criterion(output, y_true)

    def acc(y_pred, y_true, *args):
        y_pred, _ = y_pred
        return metrics.bin_acc(y_pred, y_true, *args)
```

Which is totally normal considering that this is an uncommon output and there is no way to know how to calculate the loss or the metric in a generic way. Training was working just fine with this. But when it came to `predict()` it couldn't get through because of 

https://github.com/GRAAL-Research/poutyne/blob/master/poutyne/framework/model.py#L406

So I experimented a little bit and came up with this function in `utils.py`
```python
def _concat(obj):
    if isinstance(obj[0], tuple):
        return tuple([_concat(ele) for ele in zip(*obj)])
    else:
        return np.concatenate(obj)
```

Then in `predict()` I call it instead of `np.concatenate` directly
```python
        pred_y = self.predict_generator(generator)
        return utils._concat(pred_y)
```

This will work as long as the model returns either a tuple or multiple values at once
`return (output1, output2)` or `return output1, output2`.

I don't know if it's a reasonable expectation. All the tests are passing. Since it requires custom loss functions (and metrics), there doesn't seem to be any new tests to add.

Just wanted your thoughts on this.

Thanks

**Comment by freud14** [comment]
Hi,
I think it's a good idea and that we should take it further than that. We should do it for any base datatypes (tuple, list, dict) like we do in the `_apply()` function. In this function, we could  also verify if the type of `obj[0]` is a numpy array (`np.ndarray`). If not just return the list of objects.

For instance, say that the network outputs 2 custom objects of class A and B respectively. Then, the function would take the list `[(A1, B1), (A2, B2), (A3, B3)]`and return the following tuple: `([A1, A2, A3], [B1, B2, B3])`. As another example, say that the network outputs this: `[{'a': ...,'b':...}, {'a': ...,'b':...}, {'a': ...,'b':...}]` where `...` are numpy arrays (actually tensors converted into numpy arrays). Then, the function would return `{'a':..., 'b':...}` where `...` is the concatenation of the respective numpy arrays.

This way, we could also apply the function in the evaluate_generator and predict_generator without breaking too much stuff. The only assumptions that this makes is that the output of the forward method is always under the same format which is fair enough I guess.

What do you think?

Frédérik

**Comment by atremblay** [comment]
No nested structure right? Nothing like
```
[
    {'a': ..., 'b': {'c': ..., 'd': (..., ...)},
    {'a': ..., 'b': {'c': ..., 'd': (..., ...)}
]
```
and return ` {'a': ..., 'b': {'c': ..., 'd': (..., ...)}`. 

Only list of either `tuples` or `dict`? Maybe recursively go through the structure might work. I will look into it this week.

**Comment by atremblay** [comment]
Digging a bit more. I think my use case is a very specific case where a modification like this would be okay; my model has two outputs, but I still uses just one for the loss. Whereas if we think of a multi-task learning where you would actually want to compute the loss over the multiple outputs then we seem to be screwed.

https://github.com/GRAAL-Research/poutyne/blob/master/poutyne/framework/model.py#L305

The targets (except the last one) would end up being lumped with the input as if it was a multiple input and then fail. Looking at Keras they seem to be standardizing that the input is either an array or a list of array. Same thing for the output. If we go this way we'll need to revisit the multiple input it seems.

**Comment by freud14** [comment]
> Maybe recursively go through the structure might work. I will look into it this week.

Yeah, I meant recursively.

I'm not sure to understand your last comment. `y_true` can be whatever you want, for instance, it can also be a tuple like `y_pred`. If you want to have two loss functions, just make one that calls the two functions with their respective inputs. I thought that your post was about aggregating the predictions when there are multiple batches?

**Comment by atremblay** [comment]
It is indeed my intention to aggregate the predictions when there are multiple batches. My concern was about the argument (un)packing. If the data returned by the StepIterator is [x1, x2, y1, y2] and we pack them as (*x, y) then x=[x1, x2, y1] and y = y2. Since the multiple input modification, it is assumed that all items except the last one are part of the input. It made things confusing overall.

I've been experimenting a slightly different direction
https://github.com/GRAAL-Research/poutyne/compare/master...gbmarc1:multi_IO?expand=1
It would be back to `fit(self, x, y, ...)` but `x` and `y` could either be tensors/ndarray or list/tuple of tensors/ndarray. It would be closer to Keras.

**Comment by freud14** [comment]
You just have to pack them as [x1, x2, (y1, y2)], then unpacking with (*x, y) gives us x = [x1, x2] and y = (y1, y2). I don't know it is the ideal behavior.

**Comment by atremblay** [comment]
You're right, I think it would be okay this way. Writing a generator that yields `x1, x2, (y1, y2)` seems a bit inconsistent though. I'll keep doing what I started and do a PR later. Maybe we can take it from there and see if you like that direction.

**Comment by freud14** [comment]
> You're right, I think it would be okay this way. Writing a generator that yields `x1, x2, (y1, y2)` seems a bit inconsistent though.

Indeed.

> I'll keep doing what I started and do a PR later. Maybe we can take it from there and see if you like that direction. 

Yeah, sure.
